pkg/plugin/integration_test.go

```diff
@@ -351,27 +316,377 @@ func TestCreateThenUpdateCausesNoRotations(t *testing.T) {
+	blob, err := updateBlobService.Read()
+	if err != nil {
+		return nil, "", err
+	}
+	// get sync deployment checksum annotation
+	// FIXME: only doing it this way as fake kube Get on the deployment
+	// always returns an empty string on the annotation
+	syncer, err := sync.New(az.ComputeRP.Log, cs, false)
+	if err != nil {
+		return nil, "", err
+	}
+	syncChecksum, err := syncer.Hash()
+	if err != nil {
+		return nil, "", err
+	}
+	return blob, hex.EncodeToString(syncChecksum), nil
+}
+
+type rotationType string
+
+const (
+	rotationCompute rotationType = "compute"
+	rotationInfra   rotationType = "infra"
+	rotationMaster  rotationType = "master"
+	rotationSync    rotationType = "sync"
+)
+
+func getRotations(beforeBlob, afterBlob *updateblob.UpdateBlob, beforeSyncChecksum, afterSyncChecksum string) map[rotationType]bool {
+	rotations := map[rotationType]bool{rotationCompute: false, rotationInfra: false, rotationMaster: false, rotationSync: false}
+	for host := range beforeBlob.HostnameHashes {
+		rotated := !reflect.DeepEqual(beforeBlob.HostnameHashes[host], afterBlob.HostnameHashes[host])
+		if rotated {
+			rotations[rotationMaster] = true
+		}
+	}
+	for scaleset := range beforeBlob.ScalesetHashes {
+		rotated := !reflect.DeepEqual(beforeBlob.ScalesetHashes[scaleset], afterBlob.ScalesetHashes[scaleset])
+		if strings.Contains(scaleset, "compute") && rotated {
+			rotations[rotationCompute] = true
+		} else if strings.Contains(scaleset, "infra") && rotated {
+			rotations[rotationInfra] = true
+		}
+	}
+	if beforeSyncChecksum != afterSyncChecksum {
+		rotations[rotationSync] = true
+	}
+	return rotations
+}
+
+func getNodeCountFromAz(az *fakecloud.AzureCloud) map[rotationType]int {
+	nodeCount := map[rotationType]int{rotationCompute: 0, rotationInfra: 0, rotationMaster: 0}
+	for scaleset, vms := range az.ComputeRP.Vms {
+		for _, role := range []rotationType{rotationCompute, rotationInfra, rotationMaster} {
+			if strings.Contains(scaleset, string(role)) {
+				nodeCount[role] += len(vms)
+				break
+			}
+		}
+	}
+
+	return nodeCount
+}
+
+func TestHowConfigChangesCausesRotations(t *testing.T) {
+	tests := []struct {
+		name           string
+		change         func(cs *api.OpenShiftManagedCluster)
+		expectRotation map[rotationType]bool
+	}{
+		{
+			name:           "no changes",
+			expectRotation: map[rotationType]bool{rotationMaster: false, rotationInfra: false, rotationSync: false, rotationCompute: false},
+			change:         func(cs *api.OpenShiftManagedCluster) {},
+		},
+		{
+			name:           "change vm image",
+			expectRotation: map[rotationType]bool{rotationMaster: true, rotationInfra: true, rotationSync: true, rotationCompute: true},
+			change:         func(cs *api.OpenShiftManagedCluster) { cs.Config.ImageVersion = "311.12.12345678" },
+		},
+		{
+			name:           "change infra vm size",
+			expectRotation: map[rotationType]bool{rotationMaster: true, rotationInfra: true, rotationSync: false, rotationCompute: false},
```

Comment:
Why does the master rotate in this case?

Reply:
https://github.com/openshift/openshift-azure/issues/1606